src/algorithms/cf.py

```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
         return 0
     else:
         item_prob = nom/float(denom)
-        return float(item_prob)
\ No newline at end of file
+        return float(item_prob)
+
+
+def calc_naive_bayes_using_pyspark(training_data, num_partitions=20):
+    """
+    Determine the predicted rating of every user-item combination using MLlib's Naive Bayes algorithm.
+
+    Args:
+        training_data: the data used to train the RecSys algorithm in the format of a RDD of [ (userId, itemId, actualRating) ]
+
+    Returns:
+        predictions: predicted ratings of every user-item combination in the format of a RDD of [(userId, itemId, predictedRating)].
+    """
+
+    # to use MLlib's Naive Bayes model, it requires the input to be in a format of a LabeledPoint
+    # therefore, convert dataset so that it will in the format [(rating, (user, item))]
+    r_ui_train = training_data.map(lambda (u,i,r): LabeledPoint(r, (u, i)))
+    # train Naive Bayes model
+    naiveBayesModel = NaiveBayes.train(r_ui_train, lambda=1.0)
+    # predict on all user-item pairs
+    user_ids = training_data.map(lambda (u,i,r): u).distinct()
+    item_ids = training_data.map(lambda (u,i,r): i).distinct()
+    ui_combo = user_ids.cartesian(item_ids).coalesce(num_partitions)
+    r_ui_combo = ui_combo.map(lambda (u,i,r): LabeledPoint(1, (u, i)))
+    # make prediction
+    predictions = r_ui_combo.map(lambda p: (p.features[0], p.features[1], naiveBayesModel.predict(p.features)))
+
+    return predictions
+
+def __calc_naive_bayes_components(training_data, sc):
+    """
+    Helper function that will compute the necessary components needed by:
+    calc_naive_bayes_map(), calc_naive_bayes_mse(), calc_naive_bayes_mae()
+
+    """
+
+    # create RDD for range of ratings, ie. [1, 2, 3, 4, 5] for ratings 1-5
+    min_rating = training_data.map(lambda (u,i,r): r).min()
+    max_rating = training_data.map(lambda (u,i,r): r).max()
+    range_of_ratings = sc.parallelize(list(range(int(min_rating), int(max_rating + 1))))
+
+    # since we have to determine the probability of rating r for each user and item, 
+    # we have to create a RDD with [(rating, (user, item))] for each rating
+    # ie. [(rating_1, (user, item)), (rating_2, (user, item)), (rating_3, (user, item)), ..., (rating_5, (user, item))]
+    ui = training_data.map(lambda (u,i,r): (u,i))
+    rCombo_ui = rangeOfRatings.cartesian(ui).map(lambda (r, (u,i)): (float(r), (u, i)))
```

Comment:
should use range_of_ratings: rCombo_ui = rangeOfRatings.cartesian(ui).map(lambda (r, (u,i)): (float(r), (u, i)))